mlflow/models/evaluation/default_evaluator.py

```diff
@@ -1089,8 +1094,9 @@ def X(self) -> pd.DataFrame:
         """
         The features (`X`) portion of the dataset, guarded against accidental mutations.
         """
+        df = pd.DataFrame(self.dataset.features_data)
         return DefaultEvaluator._MutationGuardedData(
-            pd.DataFrame(self.dataset.features_data, columns=self.dataset.feature_names)
+            df.rename(columns=dict(zip(df.columns, self.feature_names)))
```

Comment:
Let's add comment explain why cannot use `pd.DataFrame(self.dataset.features_data, columns=self.dataset.feature_names)`, it is a pitfall.